File: Day1/E.py
res = []
'''
————单调队列————
关键点：双指针[l,r]遍历所有底边长，但是每个底边长对应的矩形的高只能是其中最矮的，故考虑维护单调递增队列
但与之前不同的是，我们仍需要底边长度信息，故不删除，而是将队列右侧大于新加数的改成等于新加数，这样想要获取区间[l,r]中最小值时直接获取l下标对应值就可以了,
而且由于要复用，右侧不用删除
'''
while True:
    rects = list(map(int,input().split()))
    if rects[0]==0:
        break
    que = []
    max_mianji = 0
    for i in range(len(rects)):
        pt = -1
        while pt>=-len(que) and que[pt]>rects[i]:
            que[pt] = rects[i]
            pt-=1
        que.append(rects[i])
        for l in range(len(que)):
            tmp = (len(que)-l)*que[l]
            if tmp>max_mianji:
                max_mianji=tmp
    res.append(max_mianji)
print(*res,sep="\n")

File: Day1/E.c
#include <stdio.h>
#include <stdlib.h>

// 函数声明
void calculateMaxArea();

int main() {
    calculateMaxArea();
    return 0;
}

void calculateMaxArea() {
    int *rects = NULL;
    int *que = NULL;
    int rectSize;
    int queSize;
    int resSize = 0;
    int *res = NULL;

    while (1) {
        // 动态分配内存用于输入
        rects = (int *)malloc(1000 * sizeof(int));
        if (rects == NULL) {
            printf("Memory allocation failed\n");
            exit(1);
        }

        // 读取一行输入并解析为整数
        rectSize = 0;
        while (scanf("%d", &rects[rectSize]) && rects[rectSize] != 0) {
            rectSize++;
        }

        if (rectSize == 0) {
            free(rects);
            break;
        }

        que = (int *)malloc(rectSize * sizeof(int));
        if (que == NULL) {
            printf("Memory allocation failed\n");
            free(rects);
            exit(1);
        }

        int max_mianji = 0;
        queSize = 0;

        for (int i = 0; i < rectSize; i++) {
            int pt = queSize - 1;
            while (pt >= 0 && que[pt] > rects[i]) {
                que[pt] = rects[i];
                pt--;
            }
            queSize = pt + 2;
            que[queSize - 1] = rects[i];
            for (int l = 0; l < queSize; l++) {
                int tmp = (queSize - l) * que[l];
                if (tmp > max_mianji) {
                    max_mianji = tmp;
                }
            }
        }

        resSize++;
        res = (int *)realloc(res, resSize * sizeof(int));
        if (res == NULL) {
            printf("Memory allocation failed\n");
            free(rects);
            free(que);
            exit(1);
        }

        res[resSize - 1] = max_mianji;
        free(rects);
        free(que);
    }

    for (int i = 0; i < resSize; i++) {
        printf("%d\n", res[i]);
    }

    free(res);
}
